File: Test 3 - Pig latin - sentence/3.py
import string

def translate_word(word):
    if not word:
        return ""

    start_punct = ""
    end_punct = ""
    core = word

    while core and core[0] in string.punctuation:
        start_punct += core[0]
        core = core[1:]
    
    # Remove pontuação do fim
    while core and core[-1] in string.punctuation:
        end_punct = core[-1] + end_punct
        core = core[:-1]

    if not core:
        return word

    if not any(c.isalpha() for c in core):
        return start_punct + core + end_punct

    is_upper = core.isupper()
    is_title = core.istitle() # Verifica se é apenas a primeira maiúscula (Ex: Stop)
    
    core_lower = core.lower()
    vowels = "aeiouy"
    translated = ""

    has_consonants = any(c not in vowels for c in core_lower)

    if not has_consonants:
        translated = core + "yay"
    else:
        vowel_index = -1
        for i, letter in enumerate(core_lower):
            if letter in vowels:
                vowel_index = i
                break
        
        if vowel_index != -1:
            prefix = core[:vowel_index]
            stem = core[vowel_index:]
            translated = stem + prefix + "ay"
        else:
            translated = core + "ay"

    if is_upper:
        translated = translated.upper()
    elif is_title:
        translated = translated.capitalize()

    return start_punct + translated + end_punct

def pig_latin(sentence):
    if not sentence:
        return ""
        
    words = sentence.split(" ")
    translated_words = [translate_word(word) for word in words]
    return " ".join(translated_words)

print(pig_latin("Stop"))
print(pig_latin("No littering"))
print(pig_latin("No shirts, no shoes, no service"))
print(pig_latin("No persons under 14 admitted"))
print(pig_latin("Hey buddy, get away from my car!"))

File: Test 3 - Pig latin - sentence/3.js
function translateWord(word) {
    if (!word) return "";

    // Separa pontuação do início e do fim usando Regex
    const match = word.match(/^([^\w\s]*)(.*?)([^\w\s]*)$/);
    
    if (!match) return word;

    const startPunct = match[1];
    const core = match[2];
    const endPunct = match[3];

    if (!core) return word;

    if (!/[a-zA-Z]/.test(core)) {
        return word;
    }

    const isUpper = core === core.toUpperCase();
    const isStrictTitle = /^[A-Z][a-z]*$/.test(core);

    const coreLower = core.toLowerCase();
    const vowels = "aeiouy";
    let translated = "";

    const hasConsonants = [...coreLower].some(c => !vowels.includes(c));

    if (!hasConsonants) {
        translated = core + "yay";
    } else {
        let vowelIndex = -1;
        for (let i = 0; i < coreLower.length; i++) {
            if (vowels.includes(coreLower[i])) {
                vowelIndex = i;
                break;
            }
        }
        
        if (vowelIndex !== -1) {
            const prefix = core.slice(0, vowelIndex);
            const stem = core.slice(vowelIndex);
            translated = stem + prefix + "ay";
        } else {
            translated = core + "ay";
        }
    }

    if (isUpper) {
        translated = translated.toUpperCase();
    } else if (isStrictTitle) {
        translated = translated.charAt(0).toUpperCase() + translated.slice(1).toLowerCase();
    }

    return startPunct + translated + endPunct;
}

function pigLatin(sentence) {
    if (!sentence) {
        return "";
    }
    
    return sentence.split(" ").map(translateWord).join(" ");
}


console.log((pigLatin("Stop")))
console.log((pigLatin("No littering")))
console.log((pigLatin("No shirts, no shoes, no service")))
console.log((pigLatin("No persons under 14 admitted")))
console.log((pigLatin("Hey buddy, get away from my car!")))